compiler/rustc_middle/src/thir.rs

```diff
@@ -408,13 +409,26 @@ pub enum ExprKind<'tcx> {
     },
     /// A literal.
     Literal {
-        literal: Const<'tcx>,
+        lit: &'tcx hir::Lit,
+        neg: bool,
+    },
+    /// For literals that don't correspond to anything in the HIR
+    ScalarLiteral {
+        lit: ty::ScalarInt,
+        user_ty: Option<Canonical<'tcx, UserType<'tcx>>>,
+    },
+    /// Associated constants and named constants
+    NamedConst {
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
         user_ty: Option<Canonical<'tcx, UserType<'tcx>>>,
-        /// The `DefId` of the `const` item this literal
-        /// was produced from, if this is not a user-written
-        /// literal value.
-        const_id: Option<DefId>,
     },
+    ConstParam {
+        literal: ty::Const<'tcx>,
```

Comment:
seems weird to use a `ty::Const` for const parameters here.

want to get back to this.

Reply:
maybe use a https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.ParamConst.html instead here?